File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/usecase/usecases/DeliverOrder.java
package com.mycompany.myshop.testkit.dsl.core.usecase.usecases;

import com.mycompany.myshop.testkit.driver.port.MyShopDriver;
import com.mycompany.myshop.testkit.driver.port.dtos.DeliverOrderRequest;
import com.mycompany.myshop.testkit.dsl.core.usecase.usecases.base.BaseMyShopUseCase;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseResult;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseContext;
import com.mycompany.myshop.testkit.dsl.core.shared.VoidVerification;

public class DeliverOrder extends BaseMyShopUseCase<Void, VoidVerification> {
    private String orderNumberResultAlias;

    public DeliverOrder(MyShopDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    public DeliverOrder orderNumber(String orderNumberResultAlias) {
        this.orderNumberResultAlias = orderNumberResultAlias;
        return this;
    }

    @Override
    public UseCaseResult<Void, VoidVerification> execute() {
        var orderNumber = context.getResultValue(orderNumberResultAlias);
        var request = DeliverOrderRequest.builder().orderNumber(orderNumber).build();
        var result = driver.deliverOrder(request).mapVoid();
        return new UseCaseResult<>(result, context, VoidVerification::new);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/usecases/DeliverOrder.ts
import type { MyShopDriver } from '../../../../driver/port/my-shop-driver.js';
import type { DeliverOrderResponse } from '../../../../driver/port/dtos/DeliverOrderResponse.js';
import { UseCaseResult } from '../../shared/use-case-result.js';
import { VoidVerification } from '../../shared/void-verification.js';
import type { UseCaseContext } from '../../shared/use-case-context.js';
import { BaseMyShopUseCase } from './base/BaseMyShopUseCase.js';

export class DeliverOrder extends BaseMyShopUseCase<DeliverOrderResponse, VoidVerification> {
  private _orderNumberResultAlias: string = '';

  constructor(driver: MyShopDriver, context: UseCaseContext) {
    super(driver, context);
  }

  orderNumber(orderNumberResultAlias: string): this {
    this._orderNumberResultAlias = orderNumberResultAlias;
    return this;
  }

  async execute(): Promise<UseCaseResult<DeliverOrderResponse, VoidVerification>> {
    const orderNumber = this.context.getResultValue(this._orderNumberResultAlias) ?? this._orderNumberResultAlias;
    const result = await this.driver.deliverOrder({ orderNumber });

    return new UseCaseResult(
      result,
      this.context,
      (_, ctx) => new VoidVerification(undefined, ctx),
    );
  }
}
